components/service/pocket/src/main/java/mozilla/components/service/pocket/ext/JSON.kt

```diff
@@ -0,0 +1,28 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.pocket.ext
+
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+
+/**
+ * Map over the JSONArray, removing null values after [transform] and ignoring data that throws a [JSONException].
+ * An empty List will be returned if all the values are null or invalid.
+ *
+ * We don't expose this function publicly because it's not generic enough yet: JSONObject should be T. We don't
+ * want to expose this temporarily because then we'll break the public API.
+ */
+internal inline fun <R : Any> JSONArray.mapObjNotNull(transform: (JSONObject) -> R?): List<R> {
```

Comment:
We have a bunch of other `org.json` extensions in `support-ktx`. Maybe we can move this one over too.

Reply:
@pocmo I chose not to move this over because then it'd have to be a public android-components API. I didn't want it public because it's not generic enough: it should be 
```kotlin
inline fun <T, R : Any> JSONArray.mapNotNull(transform: (T) -> r?): List<R> {`
```

But I had difficulty making this generic enough when I initially wrote it.

Reply:
Actually, I guess I can write something like:
```kotlin
internal inline fun <T, R : Any> JSONArray.mapNotNull(extractFun: JSONArray.(Int) -> T, transform: (T) -> R?): List<R> {
    val transformedResults = mutableListOf<R>()
    for (i in 0 until this.length()) {
        try {
            val transformed = transform(this.extractFun(i))
            if (transformed != null) { transformedResults.add(transformed) }
        } catch (e: JSONException) { /* Do nothing: we skip bad data. */ }
    }

    return transformedResults
}

fun example() {
    JSONArray().mapNotNull(JSONArray::getJSONObject) {
        4
    }
}
```

I'll move it over to the JSON module.

Reply:
@pocmo Done: you may wish to re-review this because I made the code more generic, adding risk.

Reply:
btw, another way to define this:
```kotlin
inline fun <T, R : Any> JSONArray.mapNotNull(transform: (T) -> R?): List<R> {
```

And the call site:
```kotlin
JSONArray().mapNotNull<JSONObject, Int> { it.getInt("playCount") }
```

It's theoretically less performant though because the type conversion is all casting.